tools/api-key-rotation/api_key_rotation_checker/main.py

```diff
@@ -0,0 +1,346 @@
+    request = service.projects().list()
+    try:
+        response = request.execute()
+        projects.extend(response.get("projects", []))
+    except HttpError as err:
+        logging.error(err)
+
+    return projects
+
+
+def key_analysis(projects, rotation_period):
+    """
+    Performs our rotation analysis on the available API keys.
+
+    Args:
+
+    projects - A list of GCP projects and their metadata
+    rotation_period - The rotation period in days (default is 90)
+    """
+
+    logging.info(
+        "Grabbing keys and performing analysis for a rotation periods of %s days.. \n", rotation_period) # pylint: disable = line-too-long
+
+    # Get the date x (default 90) days ago
+    rotation_date = x_days_ago(rotation_period)
+
+    # Generates an access token
+    # for our API requests
+    access_token = create_token()
+
+    # This variable is used to hold our keys depending on their creation date
+    keys_needing_rotation=[]
+    keys_not_needing_rotation=[]
+
+    # For each project, extract the project ID
+    for project in projects:
+        project_id = project["projectId"]
+        try:
+            # Use the project ID and access token to find
+            # the API keys for each project
+            apikeys = requests.get(
+                f"https://apikeys.googleapis.com/v1/projects/{project_id}/apiKeys/", # pylint: disable = line-too-long
+                params={"access_token": access_token}
+                ).json()
+        except exceptions.PermissionDenied:
+            continue
+        # If API keys exist, proceed
+        if "keys" in apikeys:
+            # Access our nested keys
+            # so we can iterate through the list
+            apikeys = apikeys["keys"]
+            # For each key in our dictionaries
+            # (API keys are dictionaries)
+            for apikey in apikeys:
+                # Google removed the "createdBy" field
+                # so only legacy keys have it
+                if "createdBy" in apikey:
+                    # Create our API key object
+                    # if it has "createdBy"
+                    key_object = ApiKey(
+                        apikey["keyId"],
+                        apikey["displayName"],
+                        apikey["createdBy"],
+                        apikey["createTime"],
+                        project_id)
+                else:
+                    # Create our API key object
+                    # if it does NOT have "createdBy"
+                    key_object = ApiKey(
+                        apikey["keyId"],
+                        apikey["displayName"],
+                        "None",
+                        apikey["createTime"],
+                        project_id)
+
+                # We need to convert
+                # our creation time for comparison
+                converted_creation_date = time_convert(key_object)
+
+                # Extract API Key ID for logging
+                key_id = key_object.key_id
```

Comment:
I have very subjective feelings here, but I would divide these loops into two / three additional functions: One would retrieve a list of keys for a project / projects. Another one would take up partitioning the keys into two groups - `keys_needing_rotation`, `keys_not_needing_rotation`. See: https://docs.python.org/3/library/itertools.html method: `partition`. 

Could you also view statistics on how many keys are all, how many keys are defective and how many keys are valid?

Reply:
> I have very subjective feelings here, but I would divide these loops into two / three additional functions: One would retrieve a list of keys for a project / projects. Another one would take up partitioning the keys into two groups - keys_needing_rotation, keys_not_needing_rotation. See: https://docs.python.org/3/library/itertools.html method: partition.

Can we move this to an issue since it is a personal preference? 

> Could you also view statistics on how many keys are all, how many keys are defective and how many keys are valid?

Yes, you can. In all honesty, there are a lot of things you could do with API keys but this specific script was something created for a customer while working in PSO. I figured it would be nice to open source for anyone interested in tracking down keys that need to be rotated. What you are suggesting sounds like a feature request and I'd be more than happy to work on it at a later date.